generators/chipyard-helpers/src/main/scala/chipyard-helpers/CodingStyle.scala

```diff
@@ -0,0 +1,53 @@
+class ExampleModule extends Module {
+  // val chisel stuff should be in snake case
+  val temp_wire = WireDefault(...)
+  val my_cool_reg = RegInit(...)
+
+  // scala global defaults should be all caps
+  val DEFAULT_VALUE = 20
+
+  // other scala vars should be in camelcase
+  val myOtherVar = 20
+
+  // scala functions that are associated with the module should be put in an obj
+  // notice how the function is also in camel case since it gives a scala value out
+  def myFoo(intBits: Int, chisel_val: UInt): Int = {
+
+  }
+
+  // notice how the function is also in snake case since it gives a chisel value out
+  def my_bar(intBits: Int, intBits2: Int): UInt = {
+
+  }
+}
+
+class ModuleTemplate {
+  // override this function to place all ios in a known location
+  // must be scaladoc commented
+  def ios: ??? = ???
+
+  // override this function to place all chisel code in a known location
+  // must be scaladoc commented
+  def implementation: ??? = ???
+}
+
+class ChiselModuleStart extends ModuleTemplate {
+  def multiplyScalaNumbers(inputNumber0: Int, inputNumber1: Int): Int = {
+    inputNumber0 * inputNumber1
+  }
+
+  def add_two_chisel_uints(in_0: UInt, in_1: UInt): UInt = {
+    in_0 + in_1
+  }
+
+  override def ios: ??? = {
```

Comment:
It seems like this is just the more restrictive old-chisel IO system

```
abstract class SimpleModule extends Module {
  val io: IO
}
```

Reply:
The point is to be more restrictive such that people can parse modules easier.

Reply:
If we want this more restrictive module to be used more commonly, then adding it to Chisel would be necessary for adoption. I think this is worth bringing up to the chisel devs at least.

Reply:
> If we want this more restrictive module to be used more commonly, then adding it to Chisel would be necessary for adoption. I think this is worth bringing up to the chisel devs at least.

I think for right now if we decide to go in this direction and if this is used internally and if this is deemed useful we can upstream it to Chisel. Right now this is so early on that I think we should wait on this. (Ditto for getting feedback, we need to establish the details a bit more before reaching out)

Reply:
> I also feel like this is a bit restrictive. What if the user wants to mix in IOs as a trait (although I think that isn't necessarily a good idea...)?

If you are mixing in IOs (doing other mixins) then this wouldn't be for you. I think of this as "training wheels" to write your Chisel module.

Reply:
Hmm so are we kind of letting the user choose either the template version vs the current way to define modules?
I guess it would be more intuitive for incoming users.

Reply:
Yea, that's the intention behind this... whether this will actually help is still yet to be decided.